charts/consul/templates/cni-daemonset.yaml

```diff
@@ -0,0 +1,91 @@
+{{- if .Values.global.imageK8s }}{{ fail "global.imageK8s is not a valid key, use global.imageK8S (note the capital 'S')" }}{{ end -}}
+{{- if (and (.Values.connectInject.cni.enabled) (not .Values.connectInject.transparentProxy.defaultEnabled)) }}{{ fail "connectInject.transparentProxy.defaultEnabled must be true if connectInject.cni.enabled is true" }}{{ end -}}
+{{- if (and .Values.connectInject.cni.enabled .Values.connectInject.transparentProxy.defaultEnabled) }}
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: {{ template "consul.fullname" . }}-cni
+  namespace: {{ .Values.connectInject.cni.namespace }}
+  labels:
+    app: {{ template "consul.name" . }}
+    chart: {{ template "consul.chart" . }}
+    heritage: {{ .Release.Service }}
+    release: {{ .Release.Name }}
+    component: cni 
+spec:
+  {{- if .Values.connectInject.cni.updateStrategy }}
+  updateStrategy:
+    {{ tpl .Values.connectInject.cni.updateStrategy . | nindent 4 | trim }}
+  {{- end }}
+  selector:
+    matchLabels:
+      app: {{ template "consul.name" . }}
+      chart: {{ template "consul.chart" . }}
+      release: {{ .Release.Name }}
+      component: cni
+  template:
+    metadata:
+      labels:
+        app: {{ template "consul.name" . }}
+        chart: {{ template "consul.chart" . }}
+        release: {{ .Release.Name }}
+        component: cni
+      annotations:
+        consul.hashicorp.com/connect-inject: "false"
+    spec:
+      # consul-cni only runs on linux operating systems
+      nodeSelector:
+        kubernetes.io/os: linux
+      tolerations:
+        # Make sure consul-cni-node gets scheduled on all nodes.
+        - effect: NoSchedule
+          operator: Exists
+        # Mark the pod as a critical add-on for rescheduling.
+        - key: CriticalAddonsOnly
+          operator: Exists
+        - effect: NoExecute
+          operator: Exists
```

Comment:
why do we need these?

Reply:
This is so that kubernetes will make an attempt to schedule the daemonset before all other pods on the server. From what I have read, Kubernetes hasn't decided yet how it wants to handle critical pod scheduling so both `CriticalAddonsOnly` and `priorityClassName: system-node-critical` are two separate half-baked attempts to set some sort of priority. Both of those settings came about while I was researching bugs that other plugins had encountered...

Reply:
I think these two settings are doing 2 different things but are very similar.

- CriticalAddonsOnly: will make sure that a pod gets scheduled before other pods. Think of a scenario where 3 pods were restarted. consul-cni will get scheduled before the other pods.
- system-node-critical: will make sure that the consul-cni daemonset is started on the node immediately on startup.

I think both cases may be handled by different schedulers/controllers.  

I don't think I can test this unless I create a big cluster with a lot of pods running and just watch which pods startup first.

Reply:
ah got it! thanks for the clarification!